WordPress/openverse-catalog
# Consolidate provider workflows using dynamic DAGs and dataclasses

## Fixes


Fixes #205 by @AetherUnbound 

## Description


**Provider Workflow DAGs:**
* Removes all of the (mostly identical/boilerplate) `<provider>_workflow.py` files in favor of dataclasses and a factory
* Some adjustments to defaults so that they reflect the most commonly used values
    * For example, most DAGs are not `dated`, are scheduled monthly, and have a 24 hr `execution_timeout`
    * Updates the execution_timeout for the following:
        * Freesound from 12 -> 24 hours (a recent run timed out after 12 hours)
        * Stocksnap from 12 -> 24 hours (recent runs approach the timeout)
* Automatically sets the `doc_md` to the docstring of the provider_api_script, if one exists 

**Ingestion DAGs:**
These are DAGs that basically generate a list of days to re-ingest data, and then run what amounts to just the `pull_data` task for each of those days. The `tsv_to_postgres_loader` DAG will actually load it into the catalog DB. Issue #548 discusses changing this to include loading in the same DAG. _This_ PR only seeks to preserve the current state of things:
* Removes all of the boilerplate `<provider>_ingestion_workflow.py` files in favor of using dataclasses and a factory
* Some updates to the `get_reingestion_day_list_list` helper to slightly clarify it and make setting defaults easier
* Test updates

FWIW I also double-confirmed that we will not lose DagRun history when these changes are applied 🎉 

## Testing Instructions

Make sure that all the DAGs load. Try running a few and make sure they write lines to disk. Optionally click through the DAG details and make sure that the appropriate execution_timeouts/etc have been set. 

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
